README.md

```diff
@@ -449,6 +450,166 @@ gometalinter.v2 --install
 gometalinter.v2 --vendor ./...
 ```
 
+### Properly refactor when fixing complexity warnings 
+Sometimes the linter shows a warning about cyclomatic complexity, e.g., for the example below you will see something like this:
+```
+warning: cyclomatic complexity 13 of function DoLotsOfThings() is high (> 10) (gocyclo)
+```
+When fixing this it is important to still stick to CLEAN Code principles like keeping functions on the same level and not just refactor just as much that the warning disappears.
+
+<details>
+ <summary>Example</summary>
+
+**Example**
+```go
+func DoLotsOfThings(input1 []someStruct, input2 *someStruct) error {
+	for _, elem := range input1 {
+		if elem.Method() == "" || elem.Route() == "" {
+			return errors.New("some error")
+		}
+
+		if !strings.HasPrefix(elem.Route(), input2.Name) {
+			return errors.New("some error")
+		}
+
+		if elem.Handler() == nil {
+			return errors.New("some error")
+		}
+
+		if input2.someFunc == nil && !elem.IsPublic() {
+			return errors.New("some error")
+		}
+
+		if input2.someOtherFunc == nil {
+			return errors.New("some error")
+		}
+
+		switch c := elem.(type) {
+		case *SomeType1:
+			c.route = strings.TrimPrefix(c.route, input2.Name)
+			c.group = input2
+		case *SomeType2:
+			c.route = strings.TrimPrefix(c.route, input2.Name)
+		default:
+			return errors.New("unknown type")
+		}
+	}
+
+	return nil
+}
+```
+
+**Don't:**
+```go
+func DoLotsOfThings(input1 []someStruct, input2 *someStruct) error {
+	for _, elem := range input1 {
+		if elem.Method() == "" || elem.Route() == "" {
+			return errors.New("some error")
+		}
+
+		if !strings.HasPrefix(elem.Route(), input2.Name) {
+			return errors.New("some error")
+		}
+
+		if elem.Handler() == nil {
+			return errors.New("some error")
+		}
+
+		if input2.someFunc == nil && !elem.IsPublic() {
+			return errors.New("some error")
+		}
+
+		if input2.someOtherFunc == nil {
+			return errors.New("some error")
+		}
+
+		err := applySettings(elem)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func applySettings(elem someStruct) error {
+	switch c := elem.(type) {
+	case *SomeType1:
+		c.route = strings.TrimPrefix(c.route, input2.Name)
+		c.group = input2
+	case *SomeType2:
+		c.route = strings.TrimPrefix(c.route, input2.Name)
+	default:
+		return errors.New("unknown type")
+	}
+	return nil
+}
+```
+
+**Do:**
+```go
+func DoLotsOfThings(input1 []someStruct, input2 *someStruct) error {
+	for _, elem := range input1 {
+		if err := validateRouteAndMethod(elem, input2); err != nil {
+			return err
+		}
+
+		if err := validateFunc(elem, input2); err != nil {
+			return err
+		}
+
+		if err := applySettings(elem); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateRouteAndMethod(elem someStruct, input2 *someStruct) error {
```

Comment:
I don't think we need these three functions to make the point – wdyt?

Reply:
true, fixed